import os

configfile: "config.json"

#Tools
fastqc_path = "fastqc"

rule all:
        input:
        	expand(config["merged_SMRTcells"]+"{sample}.fofn", sample = config["sample_names"]),
        	expand(config["cluster2"]+"{sample}_clustered.bam", sample = config["sample_names"])
        #	expand(config["mapped"]+"{sample}_mapped.bam", sample = config["sample_names"])

        	
#---------------------
# Reference genome and annotation were downloaded prior to running snakemake. 
# 
# pigeon prepare gencode.v38.annotation.gtf GRCh38.primary_assembly.genome.fa
#---------------------
#rule merge SMRTcells:
#---------------------
rule merge_SMRTcells:
	input:
		bam1 = lambda wildcards: config[wildcards.sample]["bam1"] + ".hifi.u.bam",
		bam2 = lambda wildcards: config[wildcards.sample]["bam2"] + ".hifi.u.bam"
	output:
		out_merge = (config["merged_SMRTcells"]+"{sample}.fofn")
	resources:
	  mem_gb=12,
	  runtime=12,
	  threads=1,
	  ntasks=1,
	  cores=1,
	  cpus=1
	shell:
		"""ls {input.bam1} {input.bam2} > {output.out_merge}"""
		
#---------------------
#rule isoseq cluster2:
#---------------------
rule isoseq_cluster2:
	input:
		in_merge = (config["merged_SMRTcells"]+"{sample}.fofn")
	output:
		out_cluster = (config["cluster2"]+"{sample}_clustered.bam")
	resources:
	  mem_gb=120,
	  runtime=550,
	  threads=10,
	  ntasks=10,
	  cores=10,
	  cpus=8
	shell:
		"""isoseq cluster2 {input.in_merge} {output.out_cluster}"""
		
		
# pbmm2 align --preset ISOSEQ --sort ../../cluster2/NA19-290_LBD_clustered.bam /tgen_labs/jfryer/projects/references/human/GRCh38/GRCh38.primary_assembly.genome.fa NA19-290_LBD_mapped.bam	

#---------------------
#rule pbmm2_map:
#---------------------
rule pbmm2_map:
	input:
		in_cluster = (config["cluster2"]+"{sample}_clustered.bam")
	output:
		out_map = (config["mapped"]+"{sample}_mapped.bam")
	params:
	  ref = (config["ref_fa"])
	resources:
	  mem_gb=64,
	  runtime=200,
	  threads=8,
	  ntasks=8,
	  cores=8,
	  cpus=8
	shell:
		"""pbmm2 align --preset ISOSEQ --sort {input.in_cluster} {params.ref} {output.out_map}"""
#---------------------
# End of Snakefile. Proceed to R scripts for differential expression. 